### PR TITLE
iox-#590 Move unused and quality level 2 classes to separate packages

### LINKED ISSUE #590: Improve modularity by creating architecure guidelines
## Brief feature description

iceoryx is developed in several CMake packages that aim to be independent. Current state is there are cross dependencies between each of them and no architecture guidelines have been created so far.

## Detailed information

Analysed iceoryx repository with [Lattix Architect](https://www.lattix.com/products-architecture-issues/). [Attached](https://github.com/eclipse-iceoryx/iceoryx/files/6071770/iceoryx_b929aa_Architecture.Metrics_Metrics.Architecture.xlsx) are the the dependecy structure matrix and the architecture metrixs of Lattix (commit b929aa8d5022471a318cdc8e6994f1cdd6aa3850).

### General

* [x] Analyse Lattix
  * [x] Dependecy structure matrix (DSM)
  * [x] Architectual metrics
  * [x] Conceptual architecture
* [x] Namespaces & CMake package/library refactoring
* [x] Add CMake target for quality level 5/playground features and move "maybe obsolote" constructs there
  * [x] Add script to compare used namespaces to a whitelist in binary (`nm -C myBinary -f posix` with C++filt for demangling) to make sure no experimental/playground code is baked into releases
* [ ] RouDi modularisation #539
* [x] Create architecture documentation
  * [x] PlantUML or plain ASCII art? 
  * [x] Package view
  * [x] Layer view
* [ ] Summarise guidelines and write a CI checker
  * [ ] Remove `iox::helper` namespace
* [ ] Create `iceoryx_posh_core` CMake target and address the `@todo #590` in the CMake files
* [ ] Remove unused files and move quality level 2 files to a separate CMake package

### Lattix Architect findings

* [x] ~~Move POSH specific errors from iceoryx_utils to iceoryx_posh~~ Moved to #1099 
* [x] `roudi_gtest.hpp` includes `iceoryx_posh/internal/roudi_environment/roudi_environment.hpp`
* [ ] Move `RouDiEnvironment` to test folder & namespace
* [ ] `iceoryx_hoofs` does not have a clear architecture and shall be better structured

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

* Move unused files to separate repository in the iceoryx organisaton
* Move quality level 2 classes to new package `iceoryx_dust` (Deemed useful software thingies)

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #590 (partly)

### CODE REVIEW

File: iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/platform_settings.hpp [hunk 1]
```diff
@@ -20,11 +20,6 @@
 
 namespace iox
 {
-namespace posix
-{
-class NamedPipe;
-}
```
Comment (elBoberido) [inline iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/platform_settings.hpp:26]: Can you please remove this also from the other platform

File: iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/platform_settings.hpp [hunk 2]
```diff
@@ -41,7 +36,6 @@ constexpr const char IOX_PATH_SEPARATORS[IOX_NUMBER_OF_PATH_SEPARATORS] = {'/',
 constexpr uint64_t IOX_UDS_SOCKET_MAX_MESSAGE_SIZE = 1024U;
 constexpr char IOX_UDS_SOCKET_PATH_PREFIX[] = "";
 constexpr const char IOX_LOCK_FILE_PATH_PREFIX[] = "C:\\Windows\\Temp\\";
-using IoxIpcChannelType = iox::posix::NamedPipe;
```
Comment (elBoberido) [inline iceoryx_hoofs/platform/win/include/iceoryx_hoofs/platform/platform_settings.hpp:44]: same here

File: iceoryx_posh/include/iceoryx_posh/internal/runtime/ipc_interface_base.hpp [hunk 3]
```diff
@@ -29,10 +29,13 @@
...
 #include "iceoryx_hoofs/platform/types.hpp"
 #include "iceoryx_hoofs/platform/unistd.hpp"
-#include "iceoryx_hoofs/posix_wrapper/named_pipe.hpp"
 #include "iceoryx_posh/iceoryx_posh_types.hpp"
 #include "iceoryx_posh/internal/runtime/ipc_message.hpp"
 
+#if defined(_WIN32)
+#include "iceoryx_dust/posix_wrapper/named_pipe.hpp"
+#endif
+
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/internal/runtime/ipc_interface_base.hpp:38]: And here we should include the `unix_domain_socket.hpp` for non-windows platforms

File: iceoryx_posh/include/iceoryx_posh/internal/runtime/ipc_interface_base.hpp [hunk 4]
```diff
@@ -42,8 +45,15 @@
...
 
+
 namespace iox
 {
+namespace platform
+{
+#if defined(_WIN32)
+using IoxIpcChannelType = iox::posix::NamedPipe;
+#endif
+} // namespace platform
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/internal/runtime/ipc_interface_base.hpp:56]: I think the platform can be removed when we define `IoxIpcChannelType` for non-windows platforms here

File: tools/scripts/analyse_blacklisted_namespaces_in_binary.sh [hunk 5]
```diff
@@ -23,7 +23,7 @@
 #=====================================================
 
 
-BLACKLISTED_NAMESPACE=(aux)
+BLACKLISTED_NAMESPACE=(foo)
```
Comment (elBoberido) [inline tools/scripts/analyse_blacklisted_namespaces_in_binary.sh:26]: I vote for `hypnotoad`

**Review by elBoberido — COMMENTED** [review]
Looks good. Just minor things

File: iceoryx_dust/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.16)
+
+set(IOX_VERSION_STRING "2.90.0")
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_dust/CMakeLists.txt:19]: @elBoberido If you mean `tools/scripts/iceoryx_release.sh` that should work without adaptions.

File: iceoryx_posh/include/iceoryx_posh/internal/runtime/ipc_interface_base.hpp (hunk 4 above)
Comment (mossmaurice) [reply iceoryx_posh/include/iceoryx_posh/internal/runtime/ipc_interface_base.hpp]: @elBoberido As this is a platform-based thing I would prefer to keep the namespace `platform`. This clearly states the intent.

**Review by elBoberido — DISMISSED** [review]

**Review by elBoberido — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]

**Pull request merged**